### PR TITLE
Change atscppapi::TransformationPlugin to use atscppapi::Continuation instead of TSCont.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

File: include/tscpp/api/TransformationPlugin.h [hunk 1]
```diff
@@ -97,13 +98,19 @@ class TransformationPlugin : public TransactionPlugin
...
+  /**
+   * Returns true if the transformation is paused, false otherwise.
+   */
+  bool isPaused() const;
+
+  /**
+   * Returns a reference to the Continuation to schedule to resume (unpause) the transformation.  It will be resumed when this
+   * continuation is executed.  Must not call this unless the transformation is paused.
+   */
+  Continuation &resumeCont();
```
Comment (SolidWallOfCode) [inline include/tscpp/api/TransformationPlugin.h]: Not sure I like this, the asymmetry bothers me. It seems like it should be `resume` and if necessary it should take the delay and thread pool. If `pause` hides the continuation, then so should `resume`.

Comment (ywkaras) [reply include/tscpp/api/TransformationPlugin.h]: I think your are assuming that the Continuation would be executed with the schedule() member function.  Are you sure there would never be a reason to execute it with the call() or httpSchedule() member funcs?

File: src/tscpp/api/TransformationPlugin.cc [hunk 2]
```diff
@@ -273,31 +285,44 @@ TransformationPlugin::~TransformationPlugin()
...
 int
-TransformationPlugin::resumeCallback(TSCont cont, TSEvent event, void *edata)
+TransformationPluginContinuation::_run(TSEvent event, void *edata)
 {
-  auto state = static_cast<TransformationPluginState *>(TSContDataGet(cont));
-  if (state) {
-    state->paused_ = false;
-    handleTransformationPluginRead(state->vconn_, state);
-  }
+  auto state     = TS_FIELD_TO_CLASS_PTR(TransformationPluginState, resume_cont_, this);
```
Comment (SolidWallOfCode) [inline src/tscpp/api/TransformationPlugin.cc]: Seems like a bit of a large change just to avoid a `TSContDataGet`.

Comment (ywkaras) [reply src/tscpp/api/TransformationPlugin.cc]: In an object-oriented system, you call the methods of objects, e.g. obj.foo() .  If the system is also event-driven, foo() may need to complete it's actions after one or a series of events happen.  So we need some approach to handling this.  This PR is an example of my preferred approach so far.  I could have instead made TransformationPluginContinuation a  base class of TransformationPluginState, and downcasted.  In the generated machine code, downcasting and using the TS_FIELD_TO_CLASS_PTR macro would be the same.  Ultimately it's largely an aesthetic choice.  The line above would become the downcast:

auto state = static_cast<TransformationPluginState *>(this);

I think it adds confusion to use base classes for something other than a clear "is-a" relationship.

File: include/tscpp/api/TransformationPlugin.h (hunk 1 above)
Comment (SolidWallOfCode) [reply include/tscpp/api/TransformationPlugin.h]: using `HttpSchedule` seems very unlikely, but `TSContCall` is more plausible. Thinking about that, I would be more tempted to have resume

*  Get the lock for the resume continuation and try to lock.
*  If the lock is acquired, `TSContCall`
*  If not, schedule immediate.

That's the best option and realistically I don't see a use case for a different approach.

Comment (ywkaras) [reply include/tscpp/api/TransformationPlugin.h]: OK should the resume() function take a timeout and a thread pool parameter, and should there be any defaults?

Comment (ywkaras) [reply include/tscpp/api/TransformationPlugin.h]: TSContCall calls Continuation::dispatchEvent():  https://github.com/apache/trafficserver/blob/867d9c8527117ba0ac61fc2e282741e720b78cf8/iocore/eventsystem/Continuation.cc#L28

Does this function already do exactly what you said?  Why should I want to do it twice?  Does **anyone** really know how this crap works?

File: src/tscpp/api/TransformationPlugin.cc (hunk 2 above)
Comment (SolidWallOfCode) [reply src/tscpp/api/TransformationPlugin.cc]: @ywkaras and I discussed this at length. My view is that subclassing `Continuation` and putting the back pointer there would be the best approach. That's a standard C++ style approach to this problem.

Comment (ywkaras) [reply src/tscpp/api/TransformationPlugin.cc]: Party pooper.

**Pull request merged**